### ISSUE TITLE
functools.partialmethod should look more like what it's impersonating.

### ISSUE THREAD

**Issue by smontanaro** [python/cpython#74315]
BPO | [30129](https://bugs.python.org/issue30129)
--- | :---
Nosy | @smontanaro, @bitdancer, @aaronchall, @smarie, @tirkarthi
Files | <li>[partial3.py](https://bugs.python.org/file46823/partial3.py "Uploaded as text/plain at 2017-04-21.23:43:43 by @smontanaro")</li><li>[ft.diff](https://bugs.python.org/file46824/ft.diff "Uploaded as text/plain at 2017-04-21.23:43:54 by @smontanaro")</li>

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = None
closed_at = None
created_at = <Date 2017-04-21.14:27:50.984>
labels = ['3.7', 'type-feature', 'library']
title = "functools.partialmethod should look more like what it's impersonating."
updated_at = <Date 2019-09-20.15:13:00.281>
user = 'https://github.com/smontanaro'
```

bugs.python.org fields:
```python
activity = <Date 2019-09-20.15:13:00.281>
actor = 'smarie'
assignee = 'none'
closed = False
closed_date = None
closer = None
components = ['Library (Lib)']
creation = <Date 2017-04-21.14:27:50.984>
creator = 'skip.montanaro'
dependencies = []
files = ['46823', '46824']
hgrepos = []
issue_num = 30129
keywords = ['patch']
message_count = 6.0
messages = ['292050', '292051', '292070', '292088', '329348', '352854']
nosy_count = 5.0
nosy_names = ['skip.montanaro', 'r.david.murray', 'Aaron Hall', 'smarie', 'xtreak']
pr_nums = []
priority = 'normal'
resolution = None
stage = None
status = 'open'
superseder = None
type = 'enhancement'
url = 'https://bugs.python.org/issue30129'
versions = ['Python 3.7']
```

</p></details>

**Comment by smontanaro** [comment]
I needed to create a partial method in Python 2.7, so I grabbed functools.partialmethod from a Python 3.5.2 install. For various reasons, one of the reasons I wanted this was to suck in some methods from a delegated class so they appeared in dir() and help() output on the primary class (the one containing the partialmethod objects). Suppose I have

```py
class Something:
  def meth(self, arg1, arg2):
    "meth doc"
    return arg1 + arg2
```


then in the metaclass for another class I construct an attribute (call it "mymeth") which is a partialmethod object. When I (for example), run pydoc, that other class's attribute appears as:

    mymeth = <functools.partial object>

It would be nice if it at least included the doc string from meth, something like:

    mymeth = <functools.partial object>
        meth doc

Even better would be a proper signature:

```py
    mymeth(self, arg1, arg2)
        meth doc
```


In my copy of functools.partialmethod, I inserted an extra line in \_\_get__, right after the call to partial():

```py
    results.__doc__ = self.func.__doc__
```


That helps a bit, as I can

```py
    print("mymeth doc:", inst.mymeth.__doc__)
```


and see

    mymeth doc: meth doc

displayed. That's not enough for help()/pydoc though.

I suspect the heavy lifting will have to be done in pydoc.Doc.document(). inspect.isroutine() returns False for functools.partial objects. I also see \_signature_get_partial() in inspect.py. That might be the source of the problem. When I create a partialmethod object in my little example, it actually looks like a functools.partial object, not a partialmethod object. It's not clear that this test:

```py
    if isinstance(partialmethod, functools.partialmethod):
```


in inspect.\_signature_from_callable() is testing for the correct type.

Apologies that I can't easily provide a detailed example. My Python 2.x metaclass example (where I'm smashing methods from one class into another) doesn't work in Python 3.x for some reason, the whole partialmethod thing isn't available in Python 2.x (inspect doesn't know about partialmethod or partial) and it's not really a "hello world"-sized example anyway. I'll beat on things a bit more to try and craft a workable Python 3.x example.

**Comment by bitdancer** [comment]
Yes, please do provide an example. Your final words do not make a convincing case that this is a problem in python3 :)

**Comment by smontanaro** [comment]
Yeah, sorry about that. I work in an environment where I can't "eject" any
code from my work computer. I've come up with a simple Python3 example, but
it will have to wait until I can recreate it from scratch on my home
computer.

**Comment by smontanaro** [comment]
Again, my apologies for the crappy initial bug report. Hopefully this comment and the two files I just attached demonstrate what I am getting at.

I just uploaded a stupid little example, partial3.py. Stupid, but still, it demonstrates part of how I think docstrings on these partial methods could be improved. If you run it (I'm using Python 3.6.1), note that the doc strings for the sum method (instance and class), look wrong:

Child.sum doc: None
c.sum doc: partial(func, \*args, \*\*keywords) - new function with partial application
    of the given arguments and keywords.

The file, ft.diff, includes a one-line patch to partialmethod.\_\_get__ which corrects the docstring for the instance of the Child class:

Child.sum doc: None
c.sum doc: sum doc

I haven't looked to see where the docstring of Child.sum could be set, but I believe it should be fairly straightforward for someone more familiar with this code. Also, the patch doesn't improve the output of pydoc:

```py
partial3.Child = class Child(builtins.object)
 |  Methods defined here:
 |  
 |  diff(self, arg1, arg2)
 |      diff doc
 |  
 |  sum = _method(self, arg2)
 |  
 |  
```
----------------------------------------------------------------------
 |  Data descriptors defined here:
 |  
 |  \_\_dict__
 |      dictionary for instance variables (if defined)
 |  
 |  \_\_weakref__
 |      list of weak references to the object (if defined)
